addons/mass_mailing_sms/models/mailing_mailing.py

```diff
@@ -324,6 +326,43 @@ def convert_links(self):
         res.update(super(Mailing, self - sms_mailings).convert_links())
         return res
 
+    def get_extra_characters_counts(self):
+        """Get character counts for sms widget computation of characters and sms counts.
+
+        :return: Character counts used for links, formatted as `{linkCharsCount: int, optOutLinkCharsCount: int}`.
+        """
+        # opt-out is formatted as "\nSTOP SMS : [base_url]/sms/[mailing_id]/[trace_code]"
+        # links are formatted as "[base_url]/r/[link_tracker_code]/s/[sms_id]"
+        if self:
+            self.ensure_one()
+
+        self.check_access_rights('write')
+
+        query = '''
+            SELECT COALESCE(max(id), 0) FROM sms_sms 
+            UNION ALL
+            SELECT code_length FROM (
+                SELECT COALESCE(LENGTH(code), 3) AS code_length
+                FROM link_tracker_code 
+                ORDER BY LENGTH(code) DESC LIMIT 1) ltc
+        '''
+        if not self.id:
+            query += 'UNION ALL SELECT COALESCE(max(id), 0) FROM mailing_mailing'
+
+        self.env.cr.execute(query)
+        results = self.env.cr.fetchall()
+        sms_id = results[0][0]
+        max_sms_id_length = max(len(str(sms_id)), 5)  # Assumes a mailing won't be more than 100k sms at once
+        code_length = results[1][0]
+        mailing_id_placeholder = str(self.id or results[2][0] + 1)
+
+        base_url = self.get_base_url()
+        opt_out_url = urljoin(base_url, f"sms/{mailing_id_placeholder}/{'x' * self.env['mailing.trace'].CODE_SIZE}")
+        return {
+            'linkCharsCount': len(urljoin(base_url, 'r/s/')) + code_length + 1 + max_sms_id_length,
```

Comment:
Could do like previously with a "fake" link generation (that is more 'explicit' when reading), like
```suggestion
            'linkCharsCount': len(urljoin(base_url, f'r/{'x' * code_length}/s/{'x' * max_sms_id_length'})),
```
as you prefer :)